pennylane/numpy/wrapper.py

```diff
@@ -0,0 +1,95 @@
+# Copyright 2018-2020 Xanadu Quantum Technologies Inc.
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""
+This module provides the PennyLane wrapper functions for modifying NumPy,
+such that it accepts the PennyLane :class:`~.tensor` class.
+"""
+from collections import Sequence
+import functools
+
+from autograd import numpy as _np
+
+from .tensor import tensor
+
+
+def extract_tensors(x):
+    """Iterate through an iterable, and extract any PennyLane
+    tensors that appear.
+    """
+    if isinstance(x, tensor):
+        # If the item is a tensor, return it
+        yield x
+    elif isinstance(x, Sequence) and not isinstance(x, (str, bytes)):
+        # If the item is a sequence, recursively look through its
+        # elements for tensors.
+        # NOTE: we choose to branch on Sequence here and not Iterable,
+        # as NumPy arrays are not Sequences.
+        for item in x:
+            yield from extract_tensors(item)
+
+
+def tensor_wrapper(obj):
+    """Decorator that wraps existing NumPy functions so that they return
+    a PennyLane :class:`~.tensor`.
+
+    Only if the decorated function returns an ``ndarray`` is the
+    output converted to a :class:`~.tensor`; this avoids superfluous conversion
+    of scalars and other native-Python types.
+    """
+
+    @functools.wraps(obj)
+    def _wrapped(*args, **kwargs):
+        """Wrapped NumPy function"""
+
+        tensor_kwargs = {}
+
+        if "requires_grad" in kwargs:
+            tensor_kwargs["requires_grad"] = kwargs.pop("requires_grad")
+        else:
+            tensor_args = list(extract_tensors(args))
+
+            if tensor_args:
+                # Unless the user specifies otherwise, if all tensors in the argument
+                # list are non-trainable, the output is also non-trainable.
+                # NOTE: Use of Python's ``all`` results in an infinite recursion,
+                # and I'm not sure why. Using ``np.all`` works fine.
+                tensor_kwargs["requires_grad"] = not _np.all(
+                    [not i.requires_grad for i in tensor_args]
+                )
+
+        # evaluate the original object
+        res = obj(*args, **kwargs)
+
+        if isinstance(res, _np.ndarray):
+            # only if the output of the object is a ndarray,
+            # then convert to a PennyLane tensor
+            res = tensor(res, **tensor_kwargs)

```

Comment:
Trying to understand :thinking: 
So a numpy function that returns a scalar float value won't have it's output marked as differentiable?

Reply:
Since the `tensor` class is a subclass of `ndarray`, we really only want functions that return `ndarrays` to now return `tensors`. We are really taking advantage of duck typing here --- `tensor` objects quack like arrays, so existing workflows will continue to work as follows.

Making _all_ NumPy functions return `tensor` objects I feel would be:

* At best slightly confusing, since `pennylane.numpy` would no longer follow exactly the vanilla NumPy API, users would have to include additional calls to `np.item()` or `np.asscalar()` where needed, and

* At worst would create edge cases and regressions.

Note that this logic here doesn't affect differentiability, since we made the decision in the ADR that all Python floats would be treated as differentiable by default (so the Autograd interface treats `x=2.0` as if `requires_grad=True`). This decision was made for backwards compatibility with demos such as qubit rotation, but we can re-evaluate down the line if backwards-compatibility is no longer an issue.